flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,339 @@
+import org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.embedded.EmbeddedChannel;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.ErrorResponse;
+import static org.apache.flink.runtime.io.network.netty.NettyTestUtil.verifyBufferResponseHeader;
+import static org.apache.flink.runtime.io.network.netty.NettyTestUtil.verifyErrorResponse;
+import static org.apache.flink.runtime.io.network.partition.InputChannelTestUtils.createRemoteInputChannel;
+import static org.apache.flink.runtime.io.network.partition.InputChannelTestUtils.createSingleInputGate;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests the client side message decoder.
+ */
+public class NettyMessageClientDecoderDelegateTest {
+
+	private static final int BUFFER_SIZE = 1024;
+
+	private static final int NUMBER_OF_BUFFER_RESPONSES = 5;
+
+	private static final NettyBufferPool ALLOCATOR = new NettyBufferPool(1);
+
+	private EmbeddedChannel channel;
+
+	private NetworkBufferPool networkBufferPool;
+
+	private SingleInputGate inputGate;
+
+	private InputChannelID inputChannelId;
+
+	private InputChannelID releasedInputChannelId;
+
+	@Before
+	public void setUp() throws IOException, InterruptedException {
+		CreditBasedPartitionRequestClientHandler handler = new CreditBasedPartitionRequestClientHandler();
+		networkBufferPool = new NetworkBufferPool(
+			NUMBER_OF_BUFFER_RESPONSES,
+			BUFFER_SIZE,
+			NUMBER_OF_BUFFER_RESPONSES);
+		channel = new EmbeddedChannel(new NettyMessageClientDecoderDelegate(handler));
+
+		inputGate = createSingleInputGate(1);
+		RemoteInputChannel inputChannel = createRemoteInputChannel(
+			inputGate,
+			new TestingPartitionRequestClient(),
+			networkBufferPool);
+		inputGate.assignExclusiveSegments();
+		inputChannel.requestSubpartition(0);
+		handler.addInputChannel(inputChannel);
+		inputChannelId = inputChannel.getInputChannelId();
+
+		SingleInputGate releasedInputGate = createSingleInputGate(1);
+		RemoteInputChannel releasedInputChannel = new InputChannelBuilder()
+			.setMemorySegmentProvider(networkBufferPool)
+			.buildRemoteAndSetToGate(inputGate);
+		releasedInputGate.close();
+		handler.addInputChannel(releasedInputChannel);
+		releasedInputChannelId = releasedInputChannel.getInputChannelId();
+	}
+
+	@After
+	public void tearDown() throws IOException {
+		inputGate.close();
```

Comment:
Add the condition of `if (inputGate != null)` before close to avoid above `setUp` throwing any exceptions? also for the following actions.